File: prototype/src/openadp/crypto.py
#!/usr/bin/env python3
"""
OpenADP Cryptographic Functions

This module implements Ed25519-based cryptographic operations for the OpenADP
system, including point arithmetic, compression/decompression, and key derivation.

Based on RFC 8032: https://datatracker.ietf.org/doc/html/rfc8032
"""

import hashlib
from typing import Tuple, Optional, Union
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
import nacl.public
import nacl.encoding

# Type definitions for clarity
Point2D = Tuple[int, int]  # (x, y) coordinates
Point4D = Tuple[int, int, int, int]  # Extended coordinates (X, Y, Z, T)
PointAny = Union[Point2D, Point4D]

# Convert (x, y) to the expanded form (x, y, 1, x*y)
def expand(point: Point2D) -> Point4D:
    """Convert a 2D point to extended 4D coordinates."""
    x, y = point
    return (x, y, 1, x * y % p)


def unexpand(point: Point4D) -> Point2D:
    """Convert extended 4D coordinates back to 2D point."""
    x, y, z, t = point
    z_inv = modp_inv(z)
    return (x * z_inv % p, y * z_inv % p)


def sha256(s: bytes) -> bytes:
    """Compute SHA-256 hash of input bytes."""
    return hashlib.sha256(s).digest()


# Base field Z_p
p: int = 2**255 - 19


def modp_inv(x: int, prime: int = p) -> int:
    """Compute modular inverse of x modulo prime using Fermat's little theorem."""
    return pow(x, prime - 2, prime)


# Curve constant
d: int = -121665 * modp_inv(121666) % p

# Group order
q: int = 2**252 + 27742317777372353535851937790883648493

## Point operations

# Points are represented as tuples (X, Y, Z, T) of extended
# coordinates, with x = X/Z, y = Y/Z, x*y = T/Z

def point_mul8(P: Point4D) -> Point4D:
    """Multiply a point by 8 quickly using repeated doubling."""
    P2 = point_add(P, P)
    P4 = point_add(P2, P2)
    return point_add(P4, P4)


def point_valid(P: Point4D) -> bool:
    """Check if a point is valid (non-zero and has correct order)."""
    if point_equal(P, zero_point):
        return False
    return point_equal(point_mul(q, P), zero_point)


def point_add(P: Point4D, Q: Point4D) -> Point4D:
    """Add two points in extended coordinates."""
    A = (P[1] - P[0]) * (Q[1] - Q[0]) % p
    B = (P[1] + P[0]) * (Q[1] + Q[0]) % p
    C = 2 * P[3] * Q[3] * d % p
    D = 2 * P[2] * Q[2] % p
    E, F, G, H = B - A, D - C, D + C, B + A
    return (E * F % p, G * H % p, F * G % p, E * H % p)


def point_mul(s: int, P: Point4D) -> Point4D:
    """Compute scalar multiplication: Q = s * P using double-and-add."""
    Q = zero_point
    while s > 0:
        if s & 1:
            Q = point_add(Q, P)
        P = point_add(P, P)
        s >>= 1
    return Q


def point_equal(P: Point4D, Q: Point4D) -> bool:
    """Check if two points are equal in projective coordinates."""
    # x1 / z1 == x2 / z2  <==>  x1 * z2 == x2 * z1
    if (P[0] * Q[2] - Q[0] * P[2]) % p != 0:
        return False
    if (P[1] * Q[2] - Q[1] * P[2]) % p != 0:
        return False
    return True

## Point compression functions

# Square root of -1
modp_sqrt_m1: int = pow(2, (p - 1) // 4, p)


def recover_x(y: int, sign: int) -> Optional[int]:
    """
    Compute corresponding x-coordinate from y and sign bit.
    
    Args:
        y: Y coordinate
        sign: Sign bit (0 or 1)
        
    Returns:
        X coordinate if valid, None if no valid x exists
    """
    if y >= p:
        return None
    
    x2 = (y * y - 1) * modp_inv(d * y * y + 1)
    if x2 == 0:
        if sign:
            return None
        else:
            return 0

    # Compute square root of x2
    x = pow(x2, (p + 3) // 8, p)
    if (x * x - x2) % p != 0:
        x = x * modp_sqrt_m1 % p
    if (x * x - x2) % p != 0:
        return None

    if (x & 1) != sign:
        x = p - x
    return x


# Base point
g_y: int = 4 * modp_inv(5) % p
g_x: int = recover_x(g_y, 0)
G: Point4D = expand((g_x, g_y))

# The mathematician Edwards believes angles should be measured clockwise from
# the Y axis rather than counter-clockwise from the X axis, and so he decided
# just for his Edwards curve to ignore thousands of years of mathematical
# precedence.  This is why the point corresponding to 0 is at (0, 1), rather
# than (1, 0).  Ugh...
zero_point: Point4D = (0, 1, 1, 0)  # Neutral element


def point_compress(P: Point4D) -> bytes:
    """Compress a point to 32 bytes."""
    zinv = modp_inv(P[2])
    x = P[0] * zinv % p
    y = P[1] * zinv % p
    return int.to_bytes(y | ((x & 1) << 255), 32, "little")


def point_decompress(s: bytes) -> Optional[Point4D]:
    """Decompress 32 bytes to a point, or None if invalid."""
    if len(s) != 32:
        raise Exception("Invalid input length for decompression")
    
    y = int.from_bytes(s, "little")
    sign = y >> 255
    y &= (1 << 255) - 1

    x = recover_x(y, sign)
    if x is None:
        return None
    else:
        return (x, y, 1, x * y % p)

## Private key functions

def secret_expand(secret: bytes) -> int:
    """Expand a 32-byte secret key to a scalar."""
    if len(secret) != 32:
        raise Exception("Bad size of private key")
    h = sha256(secret)
    a = int.from_bytes(h, "little")
    a &= (1 << 254) - 8
    a |= (1 << 254)
    return a


def secret_to_public(secret: bytes) -> bytes:
    """Convert a private key to its corresponding public key."""
    a = secret_expand(secret)
    return point_compress(point_mul(a, G))

# Compute candidate square root of x modulo p, with p = 3 (mod 4).
def sqrt4k3(x: int, p: int) -> int:
    """Compute square root mod p where p ≡ 3 (mod 4)."""
    return pow(x, (p + 1) // 4, p)

# Compute candidate square root of x modulo p, with p = 5 (mod 8).
def sqrt8k5(x: int, p: int) -> int:
    """Compute square root mod p where p ≡ 5 (mod 8)."""
    y = pow(x, (p + 3) // 8, p)
    # If the square root exists, it is either y or y*2^(p-1)/4.
    if (y * y) % p == x % p:
        return y
    else:
        z = pow(2, (p - 1) // 4, p)
        return (y * z) % p

# Decode a hexadecimal string representation of the integer.
def hexi(s: str) -> int:
    """Decode hexadecimal string to integer."""
    return int.from_bytes(bytes.fromhex(s), byteorder="big")

# Rotate a word x by b places to the left.
def rol(x: int, b: int) -> int:
    """Rotate a 64-bit word left by b bits."""
    return ((x << b) | (x >> (64 - b))) & (2**64 - 1)

# From little endian.
def from_le(s: bytes) -> int:
    """Convert little-endian bytes to integer."""
    return int.from_bytes(s, byteorder="little")

# Some low-ish level crypto specific to OpenADP.

def prefixed(a: bytes) -> bytes:
    """Return a 16-bit length prefixed (little-endian) byte string."""
    l = len(a)
    if len(a) >= 1 << 16:
        raise Exception("Input string too long")
    prefix = int.to_bytes(l, 2, "little")
    return prefix + a


def H(UID: bytes, DID: bytes, BID: bytes, pin: bytes) -> Point4D:
    """
    Hash function mapping input parameters to a valid curve point.
    
    Args:
        UID: User identifier
        DID: Device identifier  
        BID: Backup identifier
        pin: PIN value
        
    Returns:
        A valid point on the curve with correct order
    """
    s = sha256(prefixed(UID) + prefixed(DID) + prefixed(BID) + pin)
    y_base = int.from_bytes(s, "little")
    sign = y_base >> 255
    y_base &= ((1 << 255) - 1)
    counter = 0
    
    while True:
        y = y_base ^ counter
        x = recover_x(y, sign)
        if x is not None:
            # Force the point to be in a group of order q
            P = expand((x, y))
            P = point_mul8(P)
            if point_valid(P):
                return P
        counter += 1


def deriveEncKey(P: Point4D) -> bytes:
    """
    Derive an encryption key from a curve point.
    
    Args:
        P: Point in expanded format
        
    Returns:
        32-byte encryption key
    """
    p_compressed = point_compress(P)
    hkdf = HKDF(hashes.SHA256(), 32, b"", b"OpenADP enc_key derivation")
    return hkdf.derive(p_compressed)

## X25519 functions for Noise protocol

def x25519_generate_keypair() -> Tuple[bytes, bytes]:
    """Generates an X25519 keypair."""
    private_key = nacl.public.PrivateKey.generate()
    public_key = private_key.public_key
    return (bytes(private_key), bytes(public_key))

def x25519_public_key_from_private(private_key_bytes: bytes) -> bytes:
    """Derives the X25519 public key from a private key."""
    private_key = nacl.public.PrivateKey(private_key_bytes)
    return bytes(private_key.public_key)

def x25519_dh(private_key_bytes: bytes, public_key_bytes: bytes) -> bytes:
    """Performs X25519 Diffie-Hellman key exchange."""
    private_key = nacl.public.PrivateKey(private_key_bytes)
    public_key = nacl.public.PublicKey(public_key_bytes)
    shared_secret = private_key.diffie_hellman(public_key)
    return shared_secret

if __name__ == '__main__':

    A = point_mul(123, G)
    B = point_mul(456, G)
    assert point_equal(point_add(A, B), point_mul(123 + 456, G))
    assert unexpand(expand((123, 456))) == (123, 456)
    assert point_equal(point_mul(modp_inv(2, q), point_mul(2, G)), G)

File: ocrypt/common/crypto.go
// Package common provides shared cryptographic primitives and utilities for OpenADP.
package common

import (
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"math/big"

	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
)

// Point2D represents a 2D point with (x, y) coordinates
type Point2D struct {
	X, Y *big.Int
}

// Point4D represents extended coordinates (X, Y, Z, T)
type Point4D struct {
	X, Y, Z, T *big.Int
}

// Constants for Ed25519 curve
var (
	// Base field Z_p where p = 2^255 - 19
	P = new(big.Int)

	// Curve constant d = -121665 * inv(121666) mod p
	D = new(big.Int)

	// Group order q = 2^252 + 27742317777372353535851937790883648493
	Q = new(big.Int)

	// Base point G
	G = &Point4D{}

	// Zero point (neutral element)
	ZeroPoint = &Point4D{}

	// Square root of -1 mod p
	ModpSqrtM1 = new(big.Int)
)

func init() {
	// Initialize constants
	P.SetString("57896044618658097711785492504343953926634992332820282019728792003956564819949", 10) // 2^255 - 19
	Q.SetString("7237005577332262213973186563042994240857116359379907606001950938285454250989", 10)  // 2^252 + 27742317777372353535851937790883648493

	// Calculate d = -121665 * inv(121666) mod p
	inv121666 := new(big.Int)
	inv121666.SetInt64(121666)
	inv121666.ModInverse(inv121666, P)
	D.SetInt64(-121665)
	D.Mul(D, inv121666)
	D.Mod(D, P)

	// Calculate square root of -1 mod p
	exp := new(big.Int).Sub(P, big.NewInt(1))
	exp.Div(exp, big.NewInt(4))
	ModpSqrtM1.SetInt64(2)
	ModpSqrtM1.Exp(ModpSqrtM1, exp, P)

	// Initialize base point G
	gY := new(big.Int).SetInt64(4)
	inv5 := new(big.Int).SetInt64(5)
	inv5.ModInverse(inv5, P)
	gY.Mul(gY, inv5)
	gY.Mod(gY, P)

	gX := recoverX(gY, 0)
	if gX == nil {
		panic("Failed to recover base point X coordinate")
	}

	G = Expand(&Point2D{X: gX, Y: gY})

	// Initialize zero point (0, 1, 1, 0)
	ZeroPoint = &Point4D{
		X: big.NewInt(0),
		Y: big.NewInt(1),
		Z: big.NewInt(1),
		T: big.NewInt(0),
	}
}

// Expand converts a 2D point to extended 4D coordinates
func Expand(point *Point2D) *Point4D {
	xy := new(big.Int).Mul(point.X, point.Y)
	xy.Mod(xy, P)

	return &Point4D{
		X: new(big.Int).Set(point.X),
		Y: new(big.Int).Set(point.Y),
		Z: big.NewInt(1),
		T: xy,
	}
}

// Unexpand converts extended 4D coordinates back to 2D point
func Unexpand(point *Point4D) *Point2D {
	zInv := new(big.Int).ModInverse(point.Z, P)

	x := new(big.Int).Mul(point.X, zInv)
	x.Mod(x, P)

	y := new(big.Int).Mul(point.Y, zInv)
	y.Mod(y, P)

	return &Point2D{X: x, Y: y}
}

// Sha256Hash computes SHA-256 hash of input bytes
func Sha256Hash(data []byte) []byte {
	hash := sha256.Sum256(data)
	return hash[:]
}

// modpInv computes modular inverse of x modulo prime using Fermat's little theorem
func modpInv(x *big.Int) *big.Int {
	exp := new(big.Int).Sub(P, big.NewInt(2))
	result := new(big.Int).Exp(x, exp, P)
	return result
}

// PointAdd adds two points in extended coordinates
func PointAdd(p1, p2 *Point4D) *Point4D {
	// A = (Y1 - X1) * (Y2 - X2)
	a := new(big.Int).Sub(p1.Y, p1.X)
	temp := new(big.Int).Sub(p2.Y, p2.X)
	a.Mul(a, temp)
	a.Mod(a, P)

	// B = (Y1 + X1) * (Y2 + X2)
	b := new(big.Int).Add(p1.Y, p1.X)
	temp = new(big.Int).Add(p2.Y, p2.X)
	b.Mul(b, temp)
	b.Mod(b, P)

	// C = 2 * T1 * T2 * d
	c := new(big.Int).Mul(p1.T, p2.T)
	c.Mul(c, D)
	c.Mul(c, big.NewInt(2))
	c.Mod(c, P)

	// D = 2 * Z1 * Z2
	d := new(big.Int).Mul(p1.Z, p2.Z)
	d.Mul(d, big.NewInt(2))
	d.Mod(d, P)

	// E, F, G, H = B - A, D - C, D + C, B + A
	e := new(big.Int).Sub(b, a)
	e.Mod(e, P)

	f := new(big.Int).Sub(d, c)
	f.Mod(f, P)

	g := new(big.Int).Add(d, c)
	g.Mod(g, P)

	h := new(big.Int).Add(b, a)
	h.Mod(h, P)

	// Return (E * F, G * H, F * G, E * H)
	return &Point4D{
		X: new(big.Int).Mul(e, f).Mod(new(big.Int).Mul(e, f), P),
		Y: new(big.Int).Mul(g, h).Mod(new(big.Int).Mul(g, h), P),
		Z: new(big.Int).Mul(f, g).Mod(new(big.Int).Mul(f, g), P),
		T: new(big.Int).Mul(e, h).Mod(new(big.Int).Mul(e, h), P),
	}
}

// PointMul computes scalar multiplication: Q = s * P using double-and-add
func PointMul(s *big.Int, p *Point4D) *Point4D {
	// Convert to affine coordinates for clearer debugging
	zinv := modpInv(p.Z)
	affineX := new(big.Int).Mul(p.X, zinv)
	affineX.Mod(affineX, P)
	affineY := new(big.Int).Mul(p.Y, zinv)
	affineY.Mod(affineY, P)

	q := &Point4D{
		X: new(big.Int).Set(ZeroPoint.X),
		Y: new(big.Int).Set(ZeroPoint.Y),
		Z: new(big.Int).Set(ZeroPoint.Z),
		T: new(big.Int).Set(ZeroPoint.T),
	}

	pCopy := &Point4D{
		X: new(big.Int).Set(p.X),
		Y: new(big.Int).Set(p.Y),
		Z: new(big.Int).Set(p.Z),
		T: new(big.Int).Set(p.T),
	}

	sCopy := new(big.Int).Set(s)

	for sCopy.Sign() > 0 {
		if sCopy.Bit(0) == 1 {
			q = PointAdd(q, pCopy)
		}
		pCopy = PointAdd(pCopy, pCopy)
		sCopy.Rsh(sCopy, 1)
	}

	// Debug logging for result
	resultZinv := modpInv(q.Z)
	resultX := new(big.Int).Mul(q.X, resultZinv)
	resultX.Mod(resultX, P)
	resultY := new(big.Int).Mul(q.Y, resultZinv)
	resultY.Mod(resultY, P)

	return q
}

// PointEqual checks if two points are equal in projective coordinates
func PointEqual(p1, p2 *Point4D) bool {
	// x1 / z1 == x2 / z2  <==>  x1 * z2 == x2 * z1
	left := new(big.Int).Mul(p1.X, p2.Z)
	left.Mod(left, P)
	right := new(big.Int).Mul(p2.X, p1.Z)
	right.Mod(right, P)

	if left.Cmp(right) != 0 {
		return false
	}

	left = new(big.Int).Mul(p1.Y, p2.Z)
	left.Mod(left, P)
	right = new(big.Int).Mul(p2.Y, p1.Z)
	right.Mod(right, P)

	return left.Cmp(right) == 0
}

// recoverX computes corresponding x-coordinate from y and sign bit
func recoverX(y *big.Int, sign int) *big.Int {
	if y.Cmp(P) >= 0 {
		return nil
	}

	// x^2 = (y^2 - 1) / (d * y^2 + 1)
	y2 := new(big.Int).Mul(y, y)
	y2.Mod(y2, P)

	numerator := new(big.Int).Sub(y2, big.NewInt(1))
	numerator.Mod(numerator, P)

	denominator := new(big.Int).Mul(D, y2)
	denominator.Add(denominator, big.NewInt(1))
	denominator.Mod(denominator, P)

	denominatorInv := modpInv(denominator)
	x2 := new(big.Int).Mul(numerator, denominatorInv)
	x2.Mod(x2, P)

	if x2.Sign() == 0 {
		if sign != 0 {
			return nil
		}
		return big.NewInt(0)
	}

	// Compute square root of x2
	exp := new(big.Int).Add(P, big.NewInt(3))
	exp.Div(exp, big.NewInt(8))
	x := new(big.Int).Exp(x2, exp, P)

	// Check if x^2 == x2
	xSquared := new(big.Int).Mul(x, x)
	xSquared.Mod(xSquared, P)

	if xSquared.Cmp(x2) != 0 {
		x.Mul(x, ModpSqrtM1)
		x.Mod(x, P)
	}

	// Verify again
	xSquared = new(big.Int).Mul(x, x)
	xSquared.Mod(xSquared, P)
	if xSquared.Cmp(x2) != 0 {
		return nil
	}

	// Check sign
	if int(x.Bit(0)) != sign {
		x.Sub(P, x)
	}

	return x
}

// PointCompress compresses a point to 32 bytes
func PointCompress(p *Point4D) []byte {
	zinv := modpInv(p.Z)

	x := new(big.Int).Mul(p.X, zinv)
	x.Mod(x, P)

	y := new(big.Int).Mul(p.Y, zinv)
	y.Mod(y, P)

	// Set sign bit
	if x.Bit(0) == 1 {
		y.SetBit(y, 255, 1)
	}

	result := make([]byte, 32)
	yBytes := y.Bytes()

	// Copy bytes in little-endian format
	for i := 0; i < len(yBytes) && i < 32; i++ {
		result[i] = yBytes[len(yBytes)-1-i]
	}

	return result
}

// PointDecompress decompresses 32 bytes to a point
func PointDecompress(data []byte) (*Point4D, error) {
	if len(data) != 32 {
		return nil, errors.New("invalid input length for decompression")
	}

	// Convert from little-endian
	y := big.NewInt(0)
	for i := 0; i < 32; i++ {
		for bit := 0; bit < 8; bit++ {
			if (data[i]>>bit)&1 == 1 {
				y.SetBit(y, i*8+bit, 1)
			}
		}
	}

	sign := int(y.Bit(255))
	y.SetBit(y, 255, 0) // Clear sign bit

	x := recoverX(y, sign)
	if x == nil {
		return nil, errors.New("invalid point")
	}

	xy := new(big.Int).Mul(x, y)
	xy.Mod(xy, P)

	point := &Point4D{
		X: x,
		Y: y,
		Z: big.NewInt(1),
		T: xy,
	}

	// Validate the decompressed point
	if !IsValidPoint(point) {
		return nil, errors.New("invalid point: failed validation")
	}

	return point, nil
}

// SecretExpand expands a 32-byte secret key to a scalar
func SecretExpand(secret []byte) (*big.Int, error) {
	if len(secret) != 32 {
		return nil, errors.New("bad size of private key")
	}

	h := Sha256Hash(secret)
	a := new(big.Int).SetBytes(reverseBytes(h)) // Convert from little-endian

	// a &= (1 << 254) - 8
	mask := new(big.Int).Lsh(big.NewInt(1), 254)
	mask.Sub(mask, big.NewInt(8))
	a.And(a, mask)

	// a |= (1 << 254)
	bit254 := new(big.Int).Lsh(big.NewInt(1), 254)
	a.Or(a, bit254)

	return a, nil
}

// SecretToPublic converts a private key to its corresponding public key
func SecretToPublic(secret []byte) ([]byte, error) {
	a, err := SecretExpand(secret)
	if err != nil {
		return nil, err
	}

	publicPoint := PointMul(a, G)
	return PointCompress(publicPoint), nil
}

// prefixed returns a 16-bit length prefixed (little-endian) byte string
func prefixed(a []byte) []byte {
	l := len(a)
	if l >= 1<<16 {
		panic("Input string too long")
	}
	prefix := make([]byte, 2)
	prefix[0] = byte(l)
	prefix[1] = byte(l >> 8)
	return append(prefix, a...)
}

// pointMul8 multiplies a point by 8 (matching Python point_mul8)
func pointMul8(p *Point4D) *Point4D {
	// Multiply by 8 = 2^3, so we double 3 times
	result := PointAdd(p, p)          // 2P
	result = PointAdd(result, result) // 4P
	result = PointAdd(result, result) // 8P
	return result
}

// IsValidPoint checks if a point is valid using Ed25519 cofactor clearing
func IsValidPoint(p *Point4D) bool {
	if p == nil || p.X == nil || p.Y == nil || p.Z == nil || p.T == nil {
		return false
	}

	// Check that the point is not the zero point
	if PointEqual(p, ZeroPoint) {
		return false
	}

	// Ed25519 point validation using cofactor clearing:
	// A valid point P should satisfy: 8*P is not the zero point
	// This catches low-order points (which have order dividing 8)
	// Since Y coordinates are derived from X using the curve equation,
	// we know the point is already on the curve, so no need to verify that.
	eightP := pointMul8(p)

	// The point is valid if 8*P is not the zero point
	// (Zero point would indicate a low-order point)
	return !PointEqual(eightP, ZeroPoint)
}

// H computes the hash function H(UID, DID, BID, pin) -> Point
func H(uid, did, bid, pin []byte) *Point4D {
	// Concatenate all inputs with length prefixes (matching Python implementation)
	data := prefixed(uid)
	data = append(data, prefixed(did)...)
	data = append(data, prefixed(bid)...)
	data = append(data, pin...)

	// Hash and convert to point
	hash := Sha256Hash(data)

	// Convert hash to big integer and extract sign bit (matching Python)
	yBase := new(big.Int).SetBytes(reverseBytes(hash)) // little-endian
	sign := int(yBase.Bit(255))
	yBase.SetBit(yBase, 255, 0) // Clear sign bit

	counter := 0
	for {
		// XOR with counter to find valid point
		y := new(big.Int).Xor(yBase, big.NewInt(int64(counter)))

		x := recoverX(y, sign)
		if x != nil {
			// Force the point to be in a group of order q (multiply by 8)
			P := Expand(&Point2D{X: x, Y: y})
			P = pointMul8(P)
			if IsValidPoint(P) {
				return P
			}
		}
		counter++

		// Safety check to avoid infinite loop
		if counter > 1000 {
			break
		}
	}

	// Fallback to base point if no valid point found
	return G
}

// DeriveEncKey derives an encryption key from a point
func DeriveEncKey(p *Point4D) []byte {
	compressed := PointCompress(p)

	// Use HKDF to derive a 32-byte key
	salt := []byte("OpenADP-EncKey-v1")
	info := []byte("AES-256-GCM")

	hkdf := hkdf.New(sha256.New, compressed, salt, info)
	key := make([]byte, 32)
	hkdf.Read(key)

	return key
}

// X25519GenerateKeypair generates a X25519 keypair
func X25519GenerateKeypair() ([]byte, []byte, error) {
	private := make([]byte, 32)
	_, err := rand.Read(private)
	if err != nil {
		return nil, nil, err
	}

	public, err := curve25519.X25519(private, curve25519.Basepoint)
	if err != nil {
		return nil, nil, err
	}

	return private, public, nil
}

// X25519PublicKeyFromPrivate derives public key from private key
func X25519PublicKeyFromPrivate(privateKey []byte) ([]byte, error) {
	return curve25519.X25519(privateKey, curve25519.Basepoint)
}

// X25519DH performs Diffie-Hellman key exchange
func X25519DH(privateKey, publicKey []byte) ([]byte, error) {
	return curve25519.X25519(privateKey, publicKey)
}

// Helper function to reverse bytes for little-endian conversion
func reverseBytes(data []byte) []byte {
	result := make([]byte, len(data))
	for i := 0; i < len(data); i++ {
		result[i] = data[len(data)-1-i]
	}
	return result
}

// DeriveSecret creates a deterministic secret from input parameters.
// This ensures the same inputs always produce the same secret for key recovery.
func DeriveSecret(uid, did, bid, pin []byte) *big.Int {
	// Combine all input parameters
	combined := append(uid, did...)
	combined = append(combined, bid...)
	combined = append(combined, pin...)

	// Add a domain separator to prevent collisions with other uses
	domainSep := []byte("OPENADP_SECRET_DERIVATION_V1")
	combined = append(domainSep, combined...)

	// Hash the combined data
	hash := sha256.Sum256(combined)

	// Convert hash to big integer and reduce modulo Q
	secret := new(big.Int).SetBytes(hash[:])
	secret.Mod(secret, Q)

	// Ensure secret is not zero
	if secret.Sign() == 0 {
		secret.SetInt64(1)
	}

	return secret
}

// DeriveEncKey derives an encryption key from a point using HKDF
